### PR TITLE
Resolver: A dep is equivalent to one of the things it can resolve to.

### PR DESCRIPTION
**Pull request by Eh2406**
This is a series of small changes to the resolver, each one on its own is not worth the cherne, but somehow all together we can add a new optimization rule. The result is that the test in #6283 is no longer exponencial (still a large polynomial, cubick?) and so N can be bumped from 3 to 20. This also means that we pass with all the cases reported in #6258. Resolution is NP-Hard, so we are moving the slow cases around. To reduce the chance that we will be flooded by new bad cases I run the 4 proptests overnight, and they did not find a new exponencial case. 

I would recommend reviewing this commit by commit. As each change is pretty simple on its own, but the mixed diff is harder to follow. This is submitted as one big PR as that is @alexcrichton's preference.

A special thanks to @nex3, our conversation was important in convincing me that several of these changes would be needed even in an eventual PubGrub based system. And, the question "why would PubGrub not have a problem with #6283" was wat crystallized this optimization opportunity in my mind.

### CODE REVIEW

File: tests/testsuite/build.rs [hunk 1]
```diff
@@ -1226,8 +1226,7 @@ versions that meet the requirements `>= 1.0.1, <= 2.0.0` are: 2.0.0, 1.0.1
 
 all possible versions conflict with previously selected packages.
 
-  previously selected package `bad v2.0.1`
```
Comment (alexcrichton) [inline tests/testsuite/build.rs]: Hm is this a bug that this clause no longer mentions `bad`? Reading the new error message it seems like it's not as clear as it was before

File: src/cargo/core/resolver/context.rs [hunk 2]
```diff
@@ -52,8 +53,39 @@ pub struct Context {
...
+/// This all so stores the size of `Activations` when that version was add as an "age".
+/// This is used to speed up backtracking.
+pub type Activations =
+    im_rc::HashMap<(InternedString, SourceId, SemverCompatibility), (Summary, usize)>;
+
+/// A type that represents when cargo treats two Versions as compatible.
+/// Versions `a` and `b` are compatible if their left-most nonzero digit is the
+/// same.
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+pub enum SemverCompatibility {
```
Comment (alexcrichton) [inline src/cargo/core/resolver/context.rs]: I think this is a pretty clever way to store `Activations`, nice find!

File: src/cargo/core/resolver/mod.rs [hunk 3]
```diff
@@ -851,25 +852,71 @@ impl RemainingCandidates {
...
+        .is_path_from_to(&parent.package_id(), &jumpback_critical_id)
+    {
+        // We are a descendant of the trigger of the problem.
+        // The best generalization of this is to let things bubble up
+        // and let `jumpback_critical_id` figure this out.
+        return None;
     }
-    if a.minor != 0 {
-        return true;
+    // What parents dose that critical activation have
```
Comment (alexcrichton) [inline src/cargo/core/resolver/mod.rs]: s/dose/does/

File: tests/testsuite/build.rs (hunk 1 above)
Comment (Eh2406) [reply tests/testsuite/build.rs]: definitely a regression, I'll see what we can do...

**Comment by Eh2406** [comment]
We are not seeing reports of performance problems from Cargos users. So I don't think it is worth committing changes that make the code harder to understand. Now, while it is fresh in my head, is the best time to add comments/docs. What could be made more clear?

File: tests/testsuite/build.rs (hunk 1 above)
Comment (Eh2406) [reply tests/testsuite/build.rs]: Fixed.

**Comment by alexcrichton** [comment]
Oh well so to be clear this is just a super tricky chunk of Cargo inherently, so I don't think this is really the fault of this PR or anything. What would be nice, however, would be a textual long-form description of what's going on somewhere, perhaps maintained in one location. The code could then liberally be commented with respect to that master description, and it'd be much easier to review the word changes (to make sure they make sense) and then cross-reference the code to make sure changes in the code match changes in the words.

In that sense there's not really a function or two I think that would benefit from more docs, but more like a "if everything had 10% more documentation it'd probably make it easier". In that sense if you can think of any comments to add that'd be great!

**Comment by Eh2406** [comment]
I definitely agree that we need to make progress to getting this (and the rest of cargos internals) better documented. Someday I'd love to have a coherent overall view of the resolver (maybe something as great as [pubgrub's](https://medium.com/@nex3/pubgrub-2fb6470504f)). Several people have asked me for a ["formal specification"](https://www.reddit.com/r/rust/comments/a3sav1/2019_roadmap_more_like_a_wishlist_finish_and_ship/ebe0911/?context=3). In the meantime we can try to find more places to insert explanatory material and try to make sure we are using consistent terminology. 

We should try to add more type aliases. For example the inconsistently used terminology of "age" (added in this PR) should be a reference to the new type. That gives us a place to add a comment on how and why it is used. (I'll add a commit when I have a spare cycles.)
We should try to make more functions (even if there only used once) for similar reasons. In that way `generalize_conflicting` is better then the code inlined.

I think let's try to only merge PR's that leave the resolver better documented. Incremental progress can add up.

**Comment by Eh2406** [comment]
It is building for me locally....

**Comment by Eh2406** [comment]
Rebased on master. This time I even remembered to fetch first!

**Comment by Eh2406** [comment]
I think I've improve comments in several miner ways. Doing an updated overall design document is going to have to happen after pub/priv is working with whatever optimizations it takes to make that work.
How would you like to proceed?

**Comment by alexcrichton** [comment]
I think this is good to go, and yeah holding off a design doc and such sounds good to me!

@bors: r+

**Pull request merged**